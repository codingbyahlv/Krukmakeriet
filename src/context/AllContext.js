import { createContext, useEffect, useState } from "react";
import { collection, getDocs } from "firebase/firestore";
import { auth, db } from "../firebase-config";
import { onAuthStateChanged } from "firebase/auth";

export const AllContext = createContext();

export const AllContextProvider = ({ children }) => {
  const [courseBasket, setCourseBasket] = useState(
    JSON.parse(localStorage.getItem("courseBasket"))
  );
  const [productBasket, setProductBasket] = useState(
    JSON.parse(localStorage.getItem("productBasket"))
  );
  const [adminPermission, setAdminPermission] = useState(false);
  const [refresh, setRefresh] = useState(true);
  const usersRef = collection(db, "users");
  const [shopProductModalOpen, setShopProductModalOpen] = useState(false);
  const [selectedProduct, setSelectedProduct] = useState(null);
  const [showTopBtn, setShowTopBtn] = useState(false);
  const [addedToBasketPopupOpen, setAddedToBasketPopupOpen] = useState(false);
  const [basketAmount, setBasketAmount] = useState(0);

  if (courseBasket === null) {
    setCourseBasket([]);
  }
  if (productBasket === null) {
    setProductBasket([]);
  }

  useEffect(() => {
    localStorage.setItem("courseBasket", JSON.stringify(courseBasket));
  }, [courseBasket]);

  useEffect(() => {
    localStorage.setItem("productBasket", JSON.stringify(productBasket));
  }, [productBasket]);

  useEffect(() => {
    async function getUsers() {
      const data = await getDocs(usersRef);
      const users = data.docs.map((doc) => ({ ...doc.data() }));
      onAuthStateChanged(auth, async (user) => {
        if (!user) return;

        const isAdmin = users.find((a) => a.uid === user.uid);
        setAdminPermission(isAdmin.admin);
        localStorage.setItem("admin", isAdmin.admin);
      });
    }
    getUsers();
  }, [refresh, usersRef]);

  return (
    <AllContext.Provider
      value={{
        courseBasket,
        setCourseBasket,
        productBasket,
        setProductBasket,
        adminPermission,
        setAdminPermission,
        refresh,
        setRefresh,
        shopProductModalOpen,
        setShopProductModalOpen,
        selectedProduct,
        setSelectedProduct,
        showTopBtn,
        setShowTopBtn,
        addedToBasketPopupOpen,
        setAddedToBasketPopupOpen,
        basketAmount,
        setBasketAmount,
      }}
    >
      {children}
    </AllContext.Provider>
  );
};
