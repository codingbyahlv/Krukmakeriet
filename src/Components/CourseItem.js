import React, { useState } from "react";
import styles from "./CourseItem.module.css";
import { useContext } from "react";
import { AllContext } from "../context/AllContext";
import Popup from "./Popup.js";
import { FaTag, FaClock } from "react-icons/fa";

function CourseItem({ courseData }) {
  const { courseBasket, setCourseBasket } = useContext(AllContext);
  const [showPopup, setShowPopup] = useState(false);

  function addToBasket(course) {
    if (courseBasket === null) {
      setCourseBasket([{ ...course, amount: 1 }]);
    } else {
      const exist = courseBasket.find((item) => item.id === courseData.id);
      if (exist) {
        setCourseBasket(
          courseBasket.map((item) =>
            item.id === course.id
              ? { ...exist, amount: exist.amount + 1 }
              : item
          )
        );
      } else {
        setCourseBasket([...courseBasket, { ...course, amount: 1 }]);
      }
    }
  }

  const timeout = setTimeout(trigger, 2000);

  function trigger() {
    setShowPopup(false);
  }

  function removeModal() {
    clearTimeout(timeout);
  }

  return (
    <article className={styles.courseCardWrapper} data-testid="courseItem">
      <div className={styles.imgWrapper}>
        <img className={styles.courseImage} src={courseData.img} alt="" />
      </div>

      <div className={styles.textWrapper}>
        <div className={styles.infoWrapper}>
          <h3 className={styles.name}>{courseData.name}</h3>
          <p className={styles.date}>{courseData.details.replace("T", " ")}</p>
        </div>

        <div className={styles.iconWrapper}>
          <p className={styles.price}>
            <FaTag /> {courseData.price}:-
          </p>
          <p className={styles.length}>
            <FaClock /> {courseData.duration}min
          </p>
        </div>
        <p className={styles.info}>{courseData.info}</p>

        <button
          className={styles.button}
          onClick={() => {
            addToBasket(courseData);
            setShowPopup(true);
            removeModal();
          }}
        >
          Boka
        </button>
      </div>
      <Popup trigger={showPopup} setTrigger={setShowPopup}>
        <p>Din kurs Ã¤r nu lagd i varukorgen.</p>
      </Popup>
    </article>
  );
}

export default CourseItem;
