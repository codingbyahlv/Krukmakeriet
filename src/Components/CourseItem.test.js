import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import CourseItem from "./CourseItem";
import { AllContextProvider } from "../context/AllContext";
import { BrowserRouter } from "react-router-dom";
import renderer from "react-test-renderer";

const mockCourseData = [
  {
    name: "Testitem",
    id: "test123",
    img: "https://en.wikipedia.org/wiki/Pottery#/media/File:Potter_making_Pottery.jpg",
    price: 100,
  },
  {
    description: "second item",
    completed: false,
  },
  {
    description: "third item",
    completed: false,
  },
];
const MockCourseItem = () => {
  return (
    <BrowserRouter>
      <AllContextProvider>
        <CourseItem courseData={mockCourseData} />
      </AllContextProvider>
    </BrowserRouter>
  );
};

describe("CourseItem", () => {
  test("check if props exist as should", async () => {
    render(<MockCourseItem />);
    const routeForCourseItem = screen.getByTestId("courseItem");
    expect(routeForCourseItem).toBeInTheDocument();
  });

  test("Matches DOM Snapshot", () => {
    const domTree = renderer
      .create(
        <BrowserRouter>
          <AllContextProvider>
            <CourseItem courseData={mockCourseData} />
          </AllContextProvider>
        </BrowserRouter>
      )
      .toJSON();
    expect(domTree).toMatchSnapshot();
  });
});
