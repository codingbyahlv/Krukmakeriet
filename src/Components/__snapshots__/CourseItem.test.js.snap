// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CourseItem Matches DOM Snapshot 1`] = `
<article
  className="courseCardWrapper"
  data-testid="courseItem"
>
  <div
    className="imgWrapper"
  >
    <img
      alt=""
      className="courseImage"
    />
  </div>
  <div
    className="textWrapper"
  >
    <div
      className="infoWrapper"
    >
      <h3
        className="name"
      />
      <p
        className="date"
      />
    </div>
    <div
      className="iconWrapper"
    >
      <p
        className="price"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 252.118V48C0 21.49 21.49 0 48 0h204.118a48 48 0 0 1 33.941 14.059l211.882 211.882c18.745 18.745 18.745 49.137 0 67.882L293.823 497.941c-18.745 18.745-49.137 18.745-67.882 0L14.059 286.059A48 48 0 0 1 0 252.118zM112 64c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48z"
          />
        </svg>
           
        :-
      </p>
      <p
        className="length"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256,8C119,8,8,119,8,256S119,504,256,504,504,393,504,256,393,8,256,8Zm92.49,313h0l-20,25a16,16,0,0,1-22.49,2.5h0l-67-49.72a40,40,0,0,1-15-31.23V112a16,16,0,0,1,16-16h32a16,16,0,0,1,16,16V256l58,42.5A16,16,0,0,1,348.49,321Z"
          />
        </svg>
           
        min
      </p>
    </div>
    <p
      className="info"
    />
    <button
      className="button"
      onClick={[Function]}
    >
      Boka
    </button>
  </div>
</article>
`;
